src/Oracle/Oracle.cs

```diff
@@ -0,0 +1,428 @@
+{
+    public class Oracle : Plugin, IPersistencePlugin
+    {
+        private NEP6Wallet Wallet;
+        private string[] Nodes;
+        private TimeSpan MaxTaskTimeout;
+        private ConcurrentDictionary<ulong, OracleTask> PendingQueue;
+        private CancellationTokenSource CancelSource;
+        private int Counter;
+        private HashSetCache<ulong> FinishedCache;
+
+        private const int RefreshInterval = 1000 * 60 * 3;
+
+        private static readonly OracleHttpProtocol Https = new OracleHttpProtocol();
+
+        public override string Description => "Built-in oracle plugin";
+
+        public Oracle()
+        {
+            PendingQueue = new ConcurrentDictionary<ulong, OracleTask>();
+            FinishedCache = new HashSetCache<ulong>(1024 * 128);
+
+            RpcServerPlugin.RegisterMethods(this);
+
+            System.Timers.Timer timer = new System.Timers.Timer();
+            timer.Enabled = true;
+            timer.Interval = RefreshInterval;
+            timer.Start();
+            timer.Elapsed += new ElapsedEventHandler(OnTimer);
+        }
+
+        protected override void Configure()
+        {
+            var config = GetConfiguration();
+            Wallet = new NEP6Wallet(Combine(PluginsDirectory, nameof(Oracle), config.GetSection("Wallet").Value));
+            Nodes = config.GetSection("Nodes").GetChildren().Select(p => p.Get<string>()).ToArray();
+            MaxTaskTimeout = TimeSpan.FromMilliseconds(double.Parse(config.GetSection("MaxTaskTimeout").Value));
+            Https.Timeout = int.Parse(config.GetSection("HttpTimeout").Value);
+            Https.AllowPrivateHost = bool.Parse(config.GetSection("AllowPrivateHost").Value);
+        }
+
+        [RpcMethod]
+        public JObject SubmitOracleResponse(JArray _params)
+        {
+            ECPoint oraclePub = ECPoint.Parse(_params[0].AsString(), ECCurve.Secp256r1);
+            ulong requestId = (ulong)_params[1].AsNumber();
+            byte[] txSign = _params[2].AsString().HexToBytes();
+            byte[] msgSign = _params[3].AsString().HexToBytes();
+
+            var data = oraclePub.ToArray().Concat(BitConverter.GetBytes(requestId)).Concat(txSign).ToArray();
+            if (!Crypto.VerifySignature(data, msgSign, oraclePub)) throw new RpcException(-100, "Invalid sign");
+            if (FinishedCache.Contains(requestId)) throw new RpcException(-100, "Request has already finished");
+
+            using (SnapshotView snapshot = Blockchain.Singleton.GetSnapshot())
+            {
+                if (NativeContract.Oracle.GetRequest(snapshot, requestId) is null)
+                    throw new RpcException(-100, "Request is not found");
+                AddResponseTxSign(snapshot, requestId, oraclePub, txSign);
+            }
+            return new JObject();
+        }
+
+        [ConsoleCommand("start oracle", Category = "Oracle", Description = "Start oracle service")]
+        private void OnStart(string password)
```

Comment:
maybe it is better to hide the password with "***", instead of show it on the console command.

Reply:
Done